### ISSUE TITLE
Pylance returns symbols for previous version of the document

### ISSUE THREAD

**Issue by amunger** [microsoft/pylance-release#4810]
Type: <b>Bug</b>

1. create a python file with multiple symbols

```
var1 = 1
var2 = 2
```

2. delete a symbol and quickly open the goto symbols quick pick
:bug: old symbols are shown

I've confirmed that the extension host has the correct version of the document, so Pylance should be providing symbols for that version.

this only repros on my dev box, not my surface laptop.

Extension version: 2023.9.10
VS Code version: Code - Insiders 1.82.0-insider (16e9205452dd417e32f164325b589b94892846b4, 2023-09-06T08:16:52.204Z)
OS version: Windows_NT x64 10.0.22000
Modes:

<details>
<summary>System Info</summary>

|Item|Value|
|---|---|
|CPUs|Intel(R) Xeon(R) CPU E5-1620 v4 @ 3.50GHz (8 x 3492)|
|GPU Status|2d_canvas: unavailable_software<br>canvas_oop_rasterization: disabled_off<br>direct_rendering_display_compositor: disabled_off_ok<br>gpu_compositing: disabled_software<br>multiple_raster_threads: enabled_on<br>opengl: disabled_off<br>rasterization: disabled_software<br>raw_draw: disabled_off_ok<br>video_decode: disabled_software<br>video_encode: disabled_software<br>vulkan: disabled_off<br>webgl: unavailable_software<br>webgl2: unavailable_software<br>webgpu: unavailable_software|
|Load (avg)|undefined|
|Memory (System)|31.92GB (6.83GB free)|
|Process Argv|--crash-reporter-id 99cd116c-59d9-4cd2-bda6-71a895fc8076|
|Screen Reader|no|
|VM|0%|
</details><details>
<summary>A/B Experiments</summary>

```
vsliv695:30137379
vsins829:30139715
vsliv368:30146709
vsreu685:30147344
python383:30185418
vspor879:30202332
vspor708:30202333
vspor363:30204092
vslsvsres303:30308271
pythontb:30258533
pythonptprofiler:30281269
vshan820:30294714
vscod805:30301674
bridge0708:30335490
bridge0723:30353136
vsaa593:30376534
pythonvs932:30404738
py29gd2263:30784851
vsclangdf:30492506
c4g48928:30535728
dsvsc012cf:30540253
pynewext54:30618038
vscrp:30624060
a9j8j154:30646983
showlangstatbar:30737417
962ge761:30823813
ecj1e332:30687743
pythonfmttext:30716741
fixshowwlkth:30771523
showindicator:30805243
pythongtdpath:30726887
i26e3531:30792625
welcomedialog:30812478
pythonnosmt12:30779711
pythonidxpt:30768918
pythonnoceb:30776497
copilotsettingt:30808721
dsvsc013:30777762
dsvsc014:30777825
diffeditorv2:30786206
pythonlinttype:30823781
pythonmhint1:30815193
dsvsc015:30821418
pythontestfixt:30826906
pythonfb280951:30830809

```

</details>

**Comment by amunger** [comment]
![Recording 2023-09-07 at 11 46 11](https://github.com/microsoft/pylance-release/assets/2019016/dd0c5dfe-f496-4e7a-8660-d34b5a70311f)

**Comment by amunger** [comment]
Once the stale symbols are returned, we cached them for that version of the document, so the quickpick will continue to show the stale symbols until another edit is made to bump the version

**Comment by heejaechang** [comment]
so, what you are saying is when `vscode` ask `pylance` with `document symbol` request, we sometime return staled data? that's probably because we process them in BG and before we finish processing it `LSP` is asking `document symbol` and we return what we have at the moment.

**Comment by amunger** [comment]
it seems to take as long as 300 ms to update symbols for a python document with 2 lines, whereas I can't get this to repro in a typescript file with 18,000 lines and ~2300 symbols - Is that expected?

**Comment by heejaechang** [comment]
that depends, is it we taking 300ms to process the LSP request, or LSP is not sending the request? typescript use vscode API directly when we use LSP.

can you add this in your setting `"python.analysis.indexing": false` and see whether it still repro? if it is, then the issue is not what I thought.

we probably need more data to find out where things get slowed.

**Comment by amunger** [comment]
It does not repro with `"python.analysis.indexing": false` set

**Comment by heejaechang** [comment]
there are another differences between us and typescript, 

typescript has 2 servers. one that only deal with `syntax info`. `document symbol` should be one of them since it only need `parse tree` to answer the request. it should handle things like `folding range`, `smart selection` and etc. and then it has another server that deal with `semantic info` which handle more expansive requests such as `completion`, `hover` anything that requires `typing` info.

but we have only 1 server handling both, so if there is previous request that are expansive (such as inferring type of complex expression), then `document symbol` request could sit in the queue until the previous request is done (assuming it wasn't cancelled) 

but it should not happen often.

**Comment by heejaechang** [comment]
> It does not repro with "python.analysis.indexing": false set

alright then, I have fix in review. now we will check version so, for open file, we will always use up-to-date data.

**Comment by rchiodo** [comment]
This issue has been fixed in prerelease version 2023.9.12, which we've just released. You can find the changelog here: [CHANGELOG.md](https://github.com/microsoft/pylance-release/blob/main/CHANGELOG.md#2023912-8-September-2023-PreRelease)
